### ISSUE TITLE
Consider adding `WithBody` to `BaseMethodDeclarationSyntax`

### ISSUE THREAD

**Issue by ltcmelo** [dotnet/roslyn#25936]
Currently, `BaseMethodDeclarationSyntax` has a `Body` member but *not* a `WithBody` method. Even though, all of its subclasses (e.g., `MethodDeclarationSyntax`) offer this API. Is there an specific reason for such decision? If not, please consider reflecting `WithBody` in the base class too. It'd help me write generic code in a rewriter.

        public override SyntaxNode VisitDestructorDeclaration(DestructorDeclarationSyntax node)
        {
            return VisitMethodDeclarationCommon(node);
        }

        public override SyntaxNode VisitMethodDeclaration(MethodDeclarationSyntax node)
        {
            return VisitMethodDeclarationCommon(node);
        }

        private SyntaxNode VisitMethodDeclarationCommon<MethodNodeT>(MethodNodeT node)
            where MethodNodeT : BaseMethodDeclarationSyntax
        {
              // use node.WithBody() here
        }

**Comment by CyrusNajmabadi** [comment]
Tagging @mattwar .  I can't remember our reasoning here.  It does seem like it would be nice.  Note: the IDE worked around this by adding extensions like so:

https://github.com/dotnet/roslyn/blob/917945b3e72ed068491e80c93e80048926c19c74/src/Workspaces/CSharp/Portable/Extensions/BaseMethodDeclarationSyntaxExtensions.cs#L26-L41

It would be nice if this just fell out of the syntaxmodel automatically.

The way it would likely need to be done would be to have this sort of pattern:

```c#
class BaseMethodSyntaxDeclaration
{
    public BaseMethodSyntaxDeclaration WithBody(BlockSyntax body) => WithBodyCore(body);
    protected abstract BaseMethodSyntaxDeclaration WithBodyCore(BlockSyntax body);
}

class MethodDeclarationSyntax : BaseMethodDeclarationSyntax
{
    protected override BaseMethodSyntaxDeclaration WithBodyCore(BlockSyntax body) => WithBody(body);
    public new MethodDeclarationSyntax WithBody(BlockSyntax body) => // existing impl
}
```

**Comment by jcouv** [comment]
I think that would make sense. 
Tagging @jaredpar @gafter since the request affects public API.

**Comment by CyrusNajmabadi** [comment]
@jaredpar @gafter  I made a proof of concept PR for this over here: https://github.com/dotnet/roslyn/pull/25945

Interesting data:  this only adds about 60 members to our public surface area.  That's def not too bad.

It does add about 120 additional internal members to to all the forwarding and to help us simulate the covariant return types you'd want with these sorts of methods.  

IMO, that doesn't seem too bad.  

Thoughts?

**Comment by jcouv** [comment]
Cyrus merged fix for 15.8. Closing
